### PR TITLE
fix: make rsbuild listen on a random port

### LINKED ISSUE #23: Error: Port is occupied when starting from container
I tried to migrate the example to our workspace. It is a vscode editor + docker container. Then I found when executing storybook in the container, the port would be occupied before rsbuild dev server is created.

I haven't created something reproducible yet, so it might not be an issue with container environment.

Current observation:
By adding breakpoints, here
https://github.com/rspack-contrib/storybook-rsbuild/blob/d21f101d0cbe8f3e87c996fe5b7e4071c187d5ed/packages/builder-rsbuild/src/preview/virtual-module-mapping.ts#L73-L77
When `readTemplate` get executed, a browser window pops up. That's where two environment diverges. If I run storybook natively in the laptop, there is no port occupation. If I execute it in the container, after it executes, the port gets occupied, so rsbuild dev server would stop.

I'll continue investigating it. Do you have any idea what storybook does here and what is its possible cause?

### LINKED ISSUE #32: Not working on Linux
## Background

I'm trying to add storybook tests to the CI workflow, but the rsbuild crashes due to a port duplicate:

```
$ storybook dev -p 6006

=> Failed to build the preview
Error: Port "6006" is occupied, please choose another one.
    at getPort (./node_modules/.pnpm/@rsbuild+core@0.7.9/node_modules/@rsbuild/core/dist/index.cjs:2611:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getServerConfig (./node_modules/.pnpm/@rsbuild+core@0.7.9/node_modules/@rsbuild/core/dist/index.cjs:2629:20)
    at async createDevServer (./node_modules/.pnpm/@rsbuild+core@0.7.9/node_modules/@rsbuild/core/dist/index.cjs:3097:33)
    at async Module.start (./node_modules/.pnpm/storybook-builder-rsbuild@0.0.6_@rsbuild+core@0.7.9_@rspack+core@0.7.4_@swc+helpers@0.5.11__@_o2736jx4xxfyk5lxieepd5obtq/node_modules/storybook-builder-rsbuild/dist/index.js:572:25)
    at async storybookDevServer (./node_modules/.pnpm/@storybook+core-server@8.1.10_encoding@0.1.13_prettier@3.3.2_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@storybook/core-server/dist/index.js:86:8402)
    at async buildOrThrow (./node_modules/.pnpm/@storybook+core-server@8.1.10_encoding@0.1.13_prettier@3.3.2_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@storybook/core-server/dist/index.js:76:1991)
    at async buildDevStandalone (./node_modules/.pnpm/@storybook+core-server@8.1.10_encoding@0.1.13_prettier@3.3.2_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@storybook/core-server/dist/index.js:165:763)
    at async withTelemetry (./node_modules/.pnpm/@storybook+core-server@8.1.10_encoding@0.1.13_prettier@3.3.2_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/@storybook/core-server/dist/index.js:86:3935)
    at async dev (./node_modules/.pnpm/@storybook+cli@8.1.10_@babel+preset-env@7.24.7_@babel+core@7.24.7__encoding@0.1.13_react-dom@_5kyel5bv7vemys3bvbvbl2xmyu/node_modules/@storybook/cli/dist/generate.js:989:565)
```

## Reproduction steps

Here is a Dockerfile that describes how to build a problematic image:

```Dockerfile
FROM ubuntu:latest

ENV COREPACK_ENABLE_DOWNLOAD_PROMPT=0
ENV NVM_DIR=/root/.nvm

COPY . /app

WORKDIR /app

RUN apt-get update && apt-get install -y curl && \
  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh | bash && \
  . ~/.bashrc && nvm install && \
  corepack enable && pnpm i && pnpm build

EXPOSE 6006
CMD [ "bash", "-c", ". /root/.nvm/nvm.sh && pnpm run -C sandboxes/react-rsbuild storybook" ]
```

## Root causes

The following code snippet works on macOS but will throw an error on Linux:

```javascript
import net from 'node:net';

net.createServer().listen(6006, undefined); // @storybook/core
net.createServer().listen(6006, 'localhost'); // rsbuild dev server
```

https://github.com/storybookjs/storybook/blob/8505d86feeecd13a8473fa1372cf37500519d136/code/core/src/core-server/dev-server.ts#L66-L69

## Workaround

`storybook dev -p 6006 -h 0.0.0.0` can make storybook core listens to `0.0.0.0` while rsbuild still listens on `localhost` network, so that they don't running on the same hostname

### PR DESCRIPTION
**Pull request by tmkx**
close #23
close #32

### CODE REVIEW

**Comment by fi3ework** [comment]
Great work! I am wondering should Rsbuild listen to a port in middleware mode, if so, what's the purpose? 🤔 Could @chenjiahan give some guidance.

**Comment by chenjiahan** [comment]
Rsbuild server should not listen to any port in the middleware mode, @9aoy can you check this

**Review by fi3ework — APPROVED** [review]

**Comment by fi3ework** [comment]
I'm merging the PR now, we could improve the default behavior in Rsbuild after @9aoy checks the code. Thanks @tmkx!

File: packages/builder-rsbuild/src/index.ts [hunk 1]
```diff
@@ -93,10 +94,9 @@ export const start: RsbuildBuilder['start'] = async ({
       ...config,
       server: {
         ...config.server,
-        port: options.port,
-        host: 'localhost',
+        port: await getRandomPort(options.host),
+        host: options.host,
         htmlFallback: false,
-        strictPort: true,
```
Comment (9aoy) [inline packages/builder-rsbuild/src/index.ts:99]: This problem is caused by configuring `strictPort: true`. Rsbuild will parse the configuration (including calculating available port) when the server is initialized. When the port is occupied and `strictPort` is configured, an error will be reported.


image


https://github.com/web-infra-dev/rsbuild/blob/v0.7.9/packages/core/src/server/helper.ts#L219-L224

Comment (fi3ework) [reply packages/builder-rsbuild/src/index.ts:99]: Got it, I'll remove port setting.

**Comment by matheusmichels** [comment]
Hey @fi3ework, this fixed the issue when running SB in a container/remove environment. However, the HMR stopped working cause now Storybook doesn't know the dev server port.

**Comment by matheusmichels** [comment]
Hey @fi3ework, this fixed the "occupied port" issue when running Storybook in a container/remove environment. However, the HMR stopped working cause now Storybook doesn't know the dev server port.

**Comment by matheusmichels** [comment]
Hey @fi3ework, this fixed the "occupied port" issue when running Storybook from a container/remote environment. However, the HMR stopped working cause now Storybook doesn't know the dev server port.

**Comment by tmkx** [comment]
rsbuild will listen to another port due to a port conflict, but hmr still works because the rsbuild dev middleware can accept connections through the Storybook server

https://github.com/rspack-contrib/storybook-rsbuild/blob/1c539b3e80d58522126c4d37a75c22f5e32544d0/packages/builder-rsbuild/src/index.ts#L135

for a speedy verification, you can follow the instructions to confirm if it functions in a container environment:

```Dockerfile
FROM ubuntu:latest

ENV COREPACK_ENABLE_DOWNLOAD_PROMPT=0
ENV NVM_DIR=/root/.nvm

ADD https://github.com/rspack-contrib/storybook-rsbuild.git /app

WORKDIR /app

RUN apt-get update && apt-get install -y curl && \
  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh | bash && \
  . ~/.bashrc && nvm install && \
  corepack enable && pnpm i && pnpm build

EXPOSE 6006
CMD [ "bash", "-c", ". /root/.nvm/nvm.sh && pnpm -C sandboxes/react-18 exec storybook dev --no-open -p 6006" ]
```
